addons/stock/models/stock_quant.py

```diff
@@ -237,6 +237,21 @@ def check_product_id(self):
         if any(elem.product_id.type != 'product' for elem in self):
             raise ValidationError(_('Quants cannot be created for consumables or services.'))
 
+    @api.constrains('lot_id')
+    def check_lot_id(self):
+        for quant in self:
+            if quant.product_id.tracking == 'serial' and quant.lot_id and quant.quantity > 0:
+                quants = self.env['stock.quant'].search([
+                    ('id', '!=', quant.id),
+                    ('product_id', '=', quant.product_id.id),
+                    ('lot_id', '=', quant.lot_id.id),
+                    ('quantity', '>', 0),
+                    '|', ('location_id.usage', '=', 'customer'),
+                        '&', ('company_id', '=', quant.company_id.id),
+                            ('location_id.usage', 'in', ('internal', 'transit'))])
+                if quants:
+                    raise ValidationError(_("The Serial Number (%s) is already used in location(s) %s") % (quant.lot_id.name, ','.join(quants.mapped('location_id.display_name'))))
+
```

Comment:
Remove that part. Way too dangerous. It could happens strange use cases where the same sn is twice in internal location but with negative quantities that neutralize

Reply:
Indeed I was not sure that adding such constraint in stable would be a good idea
Thanks for the note